### ISSUE TITLE
The UTF8UrlDecoder can not decode the multi-byte characters

### ISSUE THREAD

**Issue by unieagle** [killbill/killbill-plugin-framework-java#8]
Hi @pierre ,

Is there any reason that we wrote our own decoder instead of using the java provided function?
And I found our version of decoder can not decode multi-byte characters.
https://github.com/killbill/killbill-plugin-framework-java/blob/master/src/main/java/org/killbill/billing/plugin/util/http/UTF8UrlDecoder.java

You can check the test as below:

```java
package org.killbill.billing.plugin.util.http;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;

import org.testng.Assert;
import org.testng.annotations.Test;

public class TestUTF8URLDecoder {

    @Test(groups = "fast")
    public void testOf() {
        verifyEncoding("abcdefghijklmnopqrstuvwxyz0123456789");
        verifyEncoding("ïłéà");
    }

    private void verifyEncoding(String input) {
        try {
            String encoded = URLEncoder.encode(input, "UTF-8");
            System.out.println(encoded);
            Assert.assertEquals(UTF8UrlDecoder.decode(encoded).toString(), URLDecoder.decode(encoded, "UTF-8"));
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
    }
}
```

The output is:

```bash
[TestNG] Running:
  /Users/unieagle/Library/Caches/IntelliJIdea15/temp-testng-customsuite.xml
abcdefghijklmnopqrstuvwxyz0123456789
%C3%AF%C5%82%C3%A9%C3%A0

java.lang.AssertionError: expected [ïłéà] but found [Ã¯ÅÃ©Ã ]
Expected :ïłéà
Actual   :Ã¯ÅÃ©Ã 

===============================================
Default Suite
Total tests run: 1, Failures: 1, Skips: 0
===============================================

Process finished with exit code 0
```

**Comment by pierre** [comment]
URLEncoder/URLDecoder only encodes and decodes form data (`application/x-www-form-urlencoded`), not URLs. Take a look at the JavaDoc of the `URL` class:

```
 * The {@link URLEncoder} and {@link URLDecoder} classes can also be
 * used, but only for HTML form encoding, which is not the same
 * as the encoding scheme defined in RFC2396.
```

Differences include how `+` is handled for instance.

That being said, there was a bug indeed in the decoding algorithm. It should be fixed by f250b111b278f549446d86ffccc6ca214a60c760.

**Comment by unieagle** [comment]
Got it, thanks for explanation!
